File: examples/python/compress_pdf.py
#!/usr/bin/env python3
import os
import sys
import time

import requests


def main() -> int:
    api_key = os.environ.get("IHATEPDF_API_KEY")
    api_base_url = os.environ.get("IHATEPDF_API_BASE_URL", "http://localhost:4000/api")
    if not api_key or len(sys.argv) != 2:
        print("Usage: IHATEPDF_API_KEY=ihp_... python3 examples/python/compress_pdf.py input.pdf", file=sys.stderr)
        return 1

    headers = {"Authorization": f"Bearer {api_key}"}
    input_path = sys.argv[1]

    with open(input_path, "rb") as source:
        upload = requests.post(
            f"{api_base_url}/v1/files",
            headers=headers,
            files={"file": (os.path.basename(input_path), source, "application/pdf")},
            timeout=60,
        )
    upload.raise_for_status()
    file_id = upload.json()["file"]["id"]

    queued = requests.post(
        f"{api_base_url}/v1/tasks/compress",
        headers={**headers, "Content-Type": "application/json"},
        json={"fileId": file_id, "outputName": "compressed.pdf"},
        timeout=30,
    )
    queued.raise_for_status()
    task = queued.json()["task"]
    print(f"Queued task: {task['id']}")

    while task["status"] not in ("completed", "failed"):
        time.sleep(1)
        status = requests.get(
            f"{api_base_url}/v1/tasks/{task['id']}/status",
            headers=headers,
            timeout=30,
        )
        status.raise_for_status()
        task = status.json()["task"]
        print(f"{task['status']} {task['progress']['percent']}%")

    if task["status"] == "failed":
        raise RuntimeError((task.get("error") or {}).get("message") or "Compression failed")

    print(task["result"]["downloadUrl"])
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/node/compress-pdf.mjs
import { readFile } from "node:fs/promises";
import { basename } from "node:path";

const apiKey = process.env.IHATEPDF_API_KEY;
const apiBaseUrl = process.env.IHATEPDF_API_BASE_URL ?? "http://localhost:4000/api";
const inputPath = process.argv[2];

if (!apiKey || !inputPath) {
  console.error("Usage: IHATEPDF_API_KEY=ihp_... node examples/node/compress-pdf.mjs input.pdf");
  process.exit(1);
}

async function jsonFetch(path, init = {}) {
  const response = await fetch(`${apiBaseUrl}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${apiKey}`,
      ...(init.headers ?? {})
    }
  });

  if (!response.ok) {
    throw new Error(`${response.status} ${await response.text()}`);
  }

  return response.json();
}

const form = new FormData();
form.set("file", new Blob([await readFile(inputPath)], { type: "application/pdf" }), basename(inputPath));

const upload = await jsonFetch("/v1/files", {
  method: "POST",
  body: form
});

const queued = await jsonFetch("/v1/tasks/compress", {
  method: "POST",
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify({
    fileId: upload.file.id,
    outputName: "compressed.pdf"
  })
});

let task = queued.task;
while (task.status !== "completed" && task.status !== "failed") {
  await new Promise((resolve) => setTimeout(resolve, 1000));
  ({ task } = await jsonFetch(`/v1/tasks/${task.id}/status`));
  console.log(`${task.status} ${task.progress.percent}%`);
}

if (task.status === "failed") {
  throw new Error(task.error?.message ?? "Compression failed");
}

console.log(task.result.downloadUrl);
